src/ucs/profile/profile.c

```diff
@@ -92,24 +110,41 @@ ucs_profile_file_write_records(int fd, ucs_profile_record_t *begin,
     return ucs_profile_file_write_data(fd, begin, (void*)end - (void*)begin);
 }
 
+/* Global lock must be held */
 static ucs_status_t
-ucs_profile_file_write_thread(int fd, ucs_time_t default_end_time)
+ucs_profile_file_write_thread(int fd, ucs_profile_thread_context_t *ctx,
+                              ucs_time_t default_end_time)
 {
     ucs_profile_thread_location_t empty_location = { .total_time = 0, .count = 0 };
     ucs_profile_thread_header_t thread_hdr;
     unsigned i, num_locations;
     ucs_status_t status;
 
+    /*
+     * NOTE: There is no protection against a race with a thread which is still
+     * producing profiling data (e.g updating the context structure without a
+     * lock).
+     * To avoid excess locking on fast-path, we assume that when we dump the
+     * profiling data (at program exit), the profiled threads are not calling
+     * ucs_profile_record() anymore.
+     */
+
+    ucs_debug("profiling context %p: write to file", ctx);
+
     /* write thread header */
     memset(&thread_hdr, 0, sizeof(thread_hdr));
```

Comment:
we don't need `memset()`
all fields are filled up below

Reply:
changes struct to packed